pkg/agent/svid/rotator.go

```diff
@@ -124,15 +130,75 @@ func (r *rotator) SetRotationFinishedHook(f func()) {
 }
 
 func (r *rotator) rotateSVIDIfNeeded(ctx context.Context) (err error) {
-	if rotationutil.ShouldRotateX509(r.clk.Now(), r.state.Value().(State).SVID[0]) {
-		err = r.rotateSVID(ctx)
-	}
-	if r.rotationFinishedHook != nil {
-		r.rotationFinishedHook()
+	state := r.state.Value().(State)
+	if rotationutil.ShouldRotateX509(r.clk.Now(), state.SVID[0]) {
+		if state.Reattestable {
+			err = r.reattest(ctx)
+		} else {
+			err = r.rotateSVID(ctx)
+		}
+
+		if r.rotationFinishedHook != nil {
+			r.rotationFinishedHook()
+		}
```

Comment:
Note i moved `rotationFinishedHook()` to only be called if an actual rotation happened. Having it called unconditionally was causing some timing issues with the unit tests. They were using this as a gate for a rotation event to be completed, but because its always being called the gate just gets passed right away and the test fails.

Reply:
Its a loop that runs every 5 seconds and calls `rotateSVIDIfNeeded()`. Perhaps this should be changed to just wake up when half ttl is reached like for workload svids?

When running the tests i see it get called more frequently than once every 5 seconds, maybe because its using mock clock?

Reply:
These tests are kinda gross and more than a little tricky so it doesn't surprise me that there is confusion here. However, the placement of the hook was intentional. It is used in tests that assert that the rotator doesn't rotate the SVID before expiration.

It is called more frequently because of the mock clock.

Reply:
The tests are a lot better with the refactoring in this PR. We're up to 77% coverage from 67%. I would consider adding https://github.com/spiffe/spire/issues/1625 to the list of issues resolved with this pr.